Some fixes, especially for dates

howff opened the pull request:
Fixes #3
* Fix the logical AND statements
* Fix the creation of integers
* Fix the recognition of dates
* Fix the output of datetimes

Commit:
Fix logical AND

Commit:
Recognise "YYYYMMDD" as datetime not numeric

Commit:
Silence warnings from pandas when guessing datetime format
and make the UK convention of DD/MM/YY

Commit:
Fix the creation of integers
(downcast doesn't truncate to int unless given ints!)

Commit:
Output datetime with matching precision
(with/without HH:MM:SS and with/without nanoseconds)

Commit:
Replace np.NaN with np.nan for numpy2 compatibility

Commit:
Replace floor('S') with floor('s') for numpy2 compatibility

Commit:
Fix for columns filled with integers that may also contain NA
cannot use astype(int) when NA present so use Int64Dtype instead
and also round floats down to remove fractional part

Commit:
Cannot use string replace to change NaN when writing out if column is int64

Commit:
Remove the inplace string replace which was causing problems

Commit:
Added sample CSV file

Commit:
Fix missing newlines

Commit:
Add a configurable limit on the number of rows read from a CSV file

IOThomas commented on BIT_+_ADR_UK_synth_data.ipynb:
```diff
@@ -832,6 +832,9 @@
         "# Replace nulls/NaNs with string\n",
         "original_data.replace(np.nan, null_string, inplace = True)\n",
+        "for column in original_data.columns:\n",
+        "    if original_data[column].dtype != pd.Int64Dtype():\n",
```
Does this mean that Int64 columns won't include the replacement strings?  I wonder if they should be converted to 'object' type at this stage so we can do that.

howff replied:
With the original code I get errors like this:
```
Traceback (most recent call last):
  File "BIT_+_ADR_UK_synth_data.py", line 591, in <module>
    original_data.replace(np.nan, null_string, inplace = True)
  File "venv/lib/python3.10/site-packages/pandas/core/generic.py", line 8141, in replace
    new_data = self._mgr.replace(
  File "venv/lib/python3.10/site-packages/pandas/core/internals/base.py", line 249, in replace
    return self.apply_with_block(
  File "venv/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 363, in apply
    applied = getattr(b, f)(**kwargs)
  File "venv/lib/python3.10/site-packages/pandas/core/internals/blocks.py", line 881, in replace
    putmask_inplace(blk.values, mask, value)
  File "venv/lib/python3.10/site-packages/pandas/core/array_algos/putmask.py", line 56, in putmask_inplace
    values[mask] = value
  File "venv/lib/python3.10/site-packages/pandas/core/arrays/masked.py", line 314, in __setitem__
    value = self._validate_setitem_value(value)
  File "venv/lib/python3.10/site-packages/pandas/core/arrays/masked.py", line 305, in _validate_setitem_value
    raise TypeError(f"Invalid value '{str(value)}' for dtype {self.dtype}")
TypeError: Invalid value '' for dtype Int64
```

I assumed the problem was that a NaN in an int column couldn't be replaced by a string because you can't put strings in an int column, but having checked it elsewhere pandas silently converts the whole column to `object` so we shouldn't see any problem. But we do. Any ideas?

howff replied:
I've tried inplace=False, assigning to a different dataframe, and I've tried doing each column separately, but same exception.

The only thing I can think of now is to use `na_rep=null_string` in `to_csv()` but that might not help with `write_sav()` as per https://github.com/BehaviouralInsights/BIT-ADRUK-synthetic-data-tool/pull/2/commits/0044d6c6db845fa7549dc95c09b50d3c94be8547

howff replied:
I've done some more tests using this code
```
import pandas as pd
import numpy as np
df = pd.DataFrame( {
    'int_pd':   [ 1,   pd.NA,   3, 4, 5],
    'int_np':   [ 1,   np.nan,  3, 4, 5],
    'float_pd': [ 1.0, pd.NA,   3, 4, 5],
    'float_np': [ 1.0, np.nan,  3, 4, 5],
    'str_pd':   [ '1', pd.NA,  '3', '4', '5'],
    'str_np':   [ '1', np.nan, '3', '4', '5'],
} )
print('-----\ndtypes before forcing int columns to Int64Dtype\n%s' % df.dtypes)
df['int_pd'] = df['int_pd'].astype(pd.Int64Dtype())
df['int_np'] = df['int_np'].astype(pd.Int64Dtype())
print('-----\ndtypes before replacing NaNs\n%s' % df.dtypes)
print('-----\nTable containing NaNs\n%s' % df)
#print(df.replace(np.nan, 'NNAA')) # TypeError: Invalid value 'NNAA' for dtype Int64
for col in df.columns:
    try:
        df[col] = df[col].astype(object).replace(np.nan, 'NANNY')
    except Exception as e:
        print('cannot replace in column %s (%s)' % (col,e))
print('-----\ndtypes after replacing NaNs\n%s' % df.dtypes)
df.to_csv('/dev/tty')
```

I've come to the conclusion that using `astype(object)` is the best solution here.
Oh, and using `inplace=True` gives warnings so assignment is used.
I've pushed a final commit https://github.com/BehaviouralInsights/BIT-ADRUK-synthetic-data-tool/pull/2/commits/e714a859801b07466d6fe440e9d6b4e73d8cc068 that I hope completes the changes.

IOThomas replied:
That looks great, thanks!

IOThomas commented on BIT_+_ADR_UK_synth_data.ipynb:
```diff
@@ -368,6 +368,8 @@
         "        return \"string\"\n",
         "    # We then check if it's numeric, or predominantly numeric with some exceptions:\n",
         "    elif(check_if_numeric(x) == True): \n",
+        "        if x.min() > 19200101 and x.max() < 20260101:",
```
This is a good catch re. a datetime format we were unaware of when writing the code!

Thinking a little about edge cases here: 

- If someone has a numerical column where all the values are between 19200101 and 20260101, then this will class it as a datetime column.  This isn't fatal, because they can redefine it as numerical using the method we provide later in the notebook, but I wonder if we can reduce the risk.  Some thoughts:
  - Screen for values that are impossible if these are date time formats  (eg. looking for digits in the month position outside of the range 0 to 12 and/or digits in the day position outside of the range 0 to 30)

howff replied:
I've pushed a commit which does what you describe.

howff replied:
Sorry, auth issues, commit now pushed https://github.com/BehaviouralInsights/BIT-ADRUK-synthetic-data-tool/pull/2/commits/28f0c30abb80b649794ec1f56afe7d84f1053248

IOThomas reviewed (commented):
Just to say again -- thanks so much for these suggestions! I think I'd be happy to incorporate them into the codebase on the whole, but I've  attached comments to two of them just to get your thoughts before I go ahead.

Commit:
Date columns like YYYYMMDD are checked for valid datetime,
and the range of dates is now configurable.
(I suppose the range check could now be removed, since we check for valid datetime?)

Commit:
Replace NaN during write to csv,
to prevent errors TypeError: Invalid value '' for dtype Int64

IOThomas commented:
Think I pressed the wrong button, so need to comment here for the responses to show up...

Commit:
Fix for replacing NaN with null_string,
convert all columns to pandas 'object' dtype.
Losing the original dtypes doesn't matter as no more operations will be performed on the columns.
This prevents the error "TypeError: Invalid value 'NNAA' for dtype Int64" when pandas can't automatically convert the dtype during the replace operation.
